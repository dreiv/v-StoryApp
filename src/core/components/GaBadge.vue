<script setup lang="ts">
import { computed, useCssModule } from 'vue'

export interface BadgeProps {
  inverted?: boolean
  information?: boolean
  error?: boolean
  warning?: boolean
  success?: boolean
  muted?: boolean
  disabled?: boolean
}

const { inverted, information, error, warning, success, muted, disabled } =
  defineProps<BadgeProps>()
const style = useCssModule()

const isDefault = computed(
  () => !inverted && !information && !error && !warning && !success && !muted && !disabled,
)

const classes = computed(() => [
  style.badge,
  { [style.default]: isDefault.value },
  { [style.inverted]: inverted },
  { [style.information]: information },
  { [style.error]: error },
  { [style.warning]: warning },
  { [style.success]: success },
  { [style.muted]: muted },
  { [style.disabled]: disabled },
])
</script>

<template>
  <span :class="classes"><slot /></span>
</template>

<style module>
.badge {
  display: inline-flex;
  justify-content: center;
  align-items: center;

  vertical-align: baseline;
  border-radius: var(--ga-radius-round);

  &:not(:empty) {
    padding: var(--ga-size-4);
    min-width: var(--ga-size-24);
    height: var(--ga-size-24);

    font-weight: var(--font-weight-medium);
    font-size: var(--text-sm);
    font-family: var(--default-font-family);

    text-align: center;
    white-space: nowrap;
  }

  &:empty {
    width: var(--ga-size-8);
    height: var(--ga-size-8);
  }
}

.default {
  background-color: var(--ga-color-surface-action);
  color: var(--ga-color-text-on-action);
}

.inverted {
  background-color: var(--ga-color-icon-on-action);
  color: var(--ga-color-text-information);
}

.muted {
  background-color: var(--ga-color-on-primary);
  color: var(--ga-color-text-action);

  &:empty {
    border: 1px solid var(--ga-color-border-disabled);
  }
}

.information {
  border: 1px solid var(--ga-color-border-information);
  background-color: var(--ga-color-utility-blue-light);
  color: var(--ga-color-text-information);

  &:empty {
    background-color: var(--ga-color-icon-information);
  }
}

.error {
  border: 1px solid var(--ga-color-border-error);
  background-color: var(--ga-color-utility-red-light);
  color: var(--ga-color-text-error);

  &:empty {
    background-color: var(--ga-color-icon-error);
  }
}

.warning {
  border: 1px solid var(--ga-color-border-warning);
  background-color: var(--ga-color-utility-orange-light);
  color: var(--ga-color-text-warning);

  &:empty {
    background-color: var(--ga-color-icon-warning);
  }
}

.success {
  border: 1px solid var(--ga-color-border-success);
  background-color: var(--ga-color-utility-green-light);
  color: var(--ga-color-text-success);

  &:empty {
    background-color: var(--ga-color-icon-success);
  }
}

.disabled {
  background-color: var(--ga-color-surface-disabled);
  color: var(--ga-color-text-disable-selected);

  &.inverted {
    background-color: var(--ga-color-utility-grey-light);
    color: var(--ga-color-text-disabled);
  }

  &:empty {
    background-color: var(--ga-color-icon-on-disabled);

    &.inverted {
      background-color: var(--ga-color-utility-grey-light);
    }
  }
}
</style>
