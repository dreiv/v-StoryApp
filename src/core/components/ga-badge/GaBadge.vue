<script setup lang="ts">
import { computed, useCssModule } from 'vue'

export interface BadgeProps {
  label?: string
  inverted?: boolean
  muted?: boolean
  information?: boolean
  error?: boolean
  warning?: boolean
  success?: boolean
  disabled?: boolean
}

const { inverted, information, error, warning, success, muted, disabled } =
  defineProps<BadgeProps>()

const style = useCssModule()
const classes = computed(() => {
  const classList = [style.badge]

  if (disabled) {
    classList.push(style.disabled)
    if (inverted) classList.push(style.inverted)
  } else if (information) {
    classList.push(style.information)
  } else if (error) {
    classList.push(style.error)
  } else if (warning) {
    classList.push(style.warning)
  } else if (success) {
    classList.push(style.success)
  } else if (muted) {
    classList.push(style.muted)
  } else if (inverted) {
    classList.push(style.inverted)
  } else {
    classList.push(style.default)
  }

  return classList
})
</script>

<template>
  <span :class="classes">
    <slot>{{ label }}</slot>
  </span>
</template>

<style module>
.badge {
  display: inline-flex;
  justify-content: center;
  align-items: center;

  border-radius: var(--ga-radius-round);

  &:not(:empty) {
    padding: var(--ga-size-spacing-02);
    min-width: var(--ga-size-spacing-06);
    height: var(--ga-size-spacing-06);

    font-weight: var(--ga-font-weight-medium);
    font-size: var(--ga-text-md-font-size);
    line-height: var(--ga-text-md-line-height);
    font-family: var(--ga-font-family-primary);

    text-align: center;
    white-space: nowrap;
  }

  &:empty {
    width: var(--ga-size-spacing-03);
    height: var(--ga-size-spacing-03);
  }
}

.default {
  background-color: var(--ga-color-surface-action);
  color: var(--ga-color-text-on-action);
}

.inverted {
  background-color: var(--ga-color-icon-on-action);
  color: var(--ga-color-text-information);
}

.muted {
  background-color: var(--ga-color-icon-on-primary);
  color: var(--ga-color-text-action);

  &:empty {
    border: var(--ga-size-border-width-sm) solid var(--ga-color-border-disabled);
  }
}

.information {
  border: var(--ga-size-border-width-sm) solid var(--ga-color-border-information);
  background-color: var(--ga-color-utility-blue-light);
  color: var(--ga-color-text-information);

  &:empty {
    background-color: var(--ga-color-icon-information);
  }
}

.error {
  border: var(--ga-size-border-width-sm) solid var(--ga-color-border-error);
  background-color: var(--ga-color-utility-red-light);
  color: var(--ga-color-text-error);

  &:empty {
    background-color: var(--ga-color-icon-error);
  }
}

.warning {
  border: var(--ga-size-border-width-sm) solid var(--ga-color-border-warning);
  background-color: var(--ga-color-utility-orange-light);
  color: var(--ga-color-text-warning);

  &:empty {
    background-color: var(--ga-color-icon-warning);
  }
}

.success {
  border: var(--ga-size-border-width-sm) solid var(--ga-color-border-success);
  background-color: var(--ga-color-utility-green-light);
  color: var(--ga-color-text-success);

  &:empty {
    background-color: var(--ga-color-icon-success);
  }
}

.disabled {
  background-color: var(--ga-color-surface-disabled);
  color: var(--ga-color-text-disable-selected);

  &.inverted {
    background-color: var(--ga-color-utility-grey-light);
    color: var(--ga-color-text-disabled);
  }

  &:not(.inverted):empty {
    background-color: var(--ga-color-icon-on-disabled);
  }
}
</style>
