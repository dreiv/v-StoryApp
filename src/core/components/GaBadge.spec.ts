import { shallowMount, type VueWrapper } from '@vue/test-utils'
import GaBadge from './GaBadge.vue'

describe('Badge', () => {
  let wrapper: VueWrapper
  beforeAll(() => {
    wrapper = shallowMount(GaBadge, { slots: { default: 'mock content' } })
  })

  it('should render correctly', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should apply the default class when no props are set', () => {
    expect(wrapper.find('.default').exists()).toBe(true)
  })

  it.each(['inverted', 'information', 'error', 'warning', 'success', 'muted', 'disabled'])(
    'should apply the correct class when the %s prop is set to true',
    async (prop) => {
      await wrapper.setProps({ [prop]: true })
      expect(wrapper.find(`.${prop}`).exists()).toBe(true)
    },
  )
})
