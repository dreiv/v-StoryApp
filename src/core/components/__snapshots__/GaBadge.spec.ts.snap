// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Badge > should render correctly with default props 1`] = `"<span class="badge default"></span>"`;

exports[`Badge > should render correctly with default slot 1`] = `"<span class="badge default">Mock Label</span>"`;

exports[`Badge > should render correctly with label prop 1`] = `"<span class="badge default">99+</span>"`;
